cql3: Enforce Key-length limits during SELECT

ptrsmrn opened the pull request:
Validate if Partition keys length is less then 64k Return InvalidRequest error according to cassandra style in case of the key is too large.

fixes #10366
fixes #12247

Minor, no need to backport.

ptrsmrn commented:
This is a continuation of https://github.com/scylladb/scylladb/pull/16012. Per https://github.com/scylladb/scylladb/pull/18407#issuecomment-2077536085, the contributor has already signed the CLA.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16336/)
:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16336/)
:x: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/4277/)
:x: - [dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/7391/)
:x: - [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/15450/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16336/)
#### Build Details:
- Duration: 6 hr 6 min
- Builder: spider7.cloudius-systems.com

nuivall commented on cql3/restrictions/statement_restrictions.cc:
```diff
@@ -1870,8 +1870,18 @@ dht::partition_range_vector partition_ranges_from_token(const expr::expression&
 }
 
+void validate_key_length(const std::vector<managed_bytes>& key) {
+    auto max_length = std::numeric_limits<compound_type<allow_prefixes::yes>::size_type>::max();
```
constexpr?

ptrsmrn replied:
No longer valid after rework, closing.

nuivall commented on test/cqlpy/cassandra_tests/validation/operations/select_test.py:
```diff
@@ -1305,7 +1304,7 @@ def testPKQueryWithValueOver64K(cql, test_keyspace):
 
 # Reproduces #10366
-@pytest.mark.xfail(reason="#10366 - server error instead of silent return")
+@pytest.mark.xfail(reason="Cassandra silently ignores oversized clustering key, Scylla chose to generate an error as in the case of an oversized partition key - see discussion in #10366")
```
`Scylla chose to generate an error` - but this test doesn't seem to check that

ptrsmrn replied:
I think these are tests @nyh rewrote 1:1 from Cassandra's repository, not sure we should change them.

nuivall replied:
yes I think you're right

nyh replied:
Yes, these tests are translated from Cassandra so we shouldn't modify except for some trivial things like slightly different error messages. 
If we decide this test actually enshrines a Cassandra **bug** and we don't want to enshrine, commenting out the test or skipping it (with a comment) is also possible. I think xfail is not the best option, because I don't think we'll ever want to be compatible with what looks like a Cassandra bug.

nuivall commented on test/cqlpy/cassandra_tests/validation/operations/select_test.py:
```diff
@@ -1296,7 +1296,6 @@ def testIndexQueryWithValueOver64K(cql, test_keyspace):
-@pytest.mark.xfail(reason="#10366 - server error instead of InvalidRequest")
 def testPKQueryWithValueOver64K(cql, test_keyspace):
     with create_table(cql, test_keyspace, "(a text, b text, PRIMARY KEY (a, b))") as table:
         TOO_BIG = 1024 * 65
```
I think we need a better test now, checking both positive and negative scenario (or 2 tests)
- MAX allowed
- MAX allowed + 1

nuivall replied:
it tests
LONG_TEXT_LENGTH = 8193
OVERSIZE_LENGTH = 66536

but not OVERSIZE_LENGTH-1

ptrsmrn replied:
If I extended `test_local_indexed_values` dtest with `OVERSIZE_LENGTH-1` case, would that be sufficient?

ptrsmrn replied:
Surprisingly, if I modify `test_local_indexed_values` to use value 65535, it errors out with:
```
E           AssertionError: Unexpected errors found:
E           node3: 1 errors
E           ERROR 2025-03-28 10:08:36,733 [shard 1:sl:d] storage_proxy - exception during mutation write to cbb7a005-8e1b-4b08-a010-e37337eb8453: std::runtime_error (Key size too large: 65537 > 65535)
```
As if we add some "metadata" (maybe type?) to the key. 
Not sure how to proceed, do we want to lower the acceptable limit by 2 then? Or more? I need to understand where this difference comes from first, I think, to make an informed decision.

ptrsmrn replied:
If I run this testcase on a pure `master` is errors out with:
```
    def testPKQueryWithValueOver64K(cql, test_keyspace):
        with create_table(cql, test_keyspace, "(a text, b text, PRIMARY KEY (a, b))") as table:
            TOO_BIG = 1024 * 65
            too_big = 'x'*TOO_BIG
>           assert_invalid_throw(cql, table, InvalidRequest,
                               "SELECT * FROM %s WHERE a = ?", too_big)

cassandra_tests/validation/operations/select_test.py:1303: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
cassandra_tests/porting.py:109: in assert_invalid_throw
    execute(cql, table, cmd, *args)
cassandra_tests/porting.py:93: in execute
    return cql.execute(prepared, args)
cassandra/cluster.py:2762: in cassandra.cluster.Session.execute
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

>   ???
E   cassandra.cluster.NoHostAvailable: ('Unable to complete the operation against any hosts', {<Host: 127.2.40.147:9042 datacenter1>: <Error from server: code=0000 [Server error] message="Key size too large: 66562 > 65535">})

ptrsmrn replied:
> it tests LONG_TEXT_LENGTH = 8193 OVERSIZE_LENGTH = 66536
> 
> but not OVERSIZE_LENGTH-1

@nuivall note that 66536 is not 64KB=65536, IDK where this number 66536 comes from.

ptrsmrn replied:
I'm a bit lost: I'm not sure if this patch is needed, seems like we're already doing the key length checks (for both the INSERT and SELECT), it's just that we print a bit different error.

ptrsmrn replied:
I added the extra testcase requested in the 1st msg.

nuivall commented on cql3/restrictions/statement_restrictions.cc:
```diff
@@ -1870,8 +1870,18 @@ dht::partition_range_vector partition_ranges_from_token(const expr::expression&
+
 /// Turns a partition-key value into a partition_range. \p pk must have elements for all partition columns.
 dht::partition_range range_from_bytes(const schema& schema, const std::vector<managed_bytes>& pk) {
+    validate_key_length(pk);
```
why check it here? it looks like a bit random place

ptrsmrn replied:
I believe it's some sort of a common place, but if you have better ideas, I'm all ears.

ptrsmrn replied:
Reworked.

ptrsmrn commented:
@nuivall thanks for the comments. I have not changed the original commit, just rebased it on top of the latest `master` and verified test.py is passing (I see 1 dtest remains to be fixed). BTW. Nadav approved the original PR with the commit I cherry-picked here. Not sure how to proceed, if I edit the commit I guess the original author won't be mentioned anymore.

ptrsmrn commented:
> @ptrsmrn - what's the next step here?

@mykaul it never got priority, it's an old issue nobody really cares about any longer.

ptrsmrn committed:
cql: return InvalidRequest for oversized partition/clustering keys

When a partition key or clustering key value exceeds the 64 KiB limit
(65535 bytes serialized), Scylla used to raise a generic
std::runtime_error "Key size too large: N > M" from the low-level
compound-key serializer. That error surfaced to clients as a CQL
server error (code 0x0000, "NoHostAvailable"-looking), which is both
ugly and incompatible with Cassandra - Cassandra returns a clean
InvalidRequest with the message "Key length of N is longer than
maximum of M".

Fix this at the single chokepoint: compound_type::serialize_value in
keys/compound.hh. The serializer is on every path that materializes a
key - INSERT/UPDATE/DELETE/BATCH build mutations through it, and
SELECT builds partition and clustering ranges through it - so a single
throw replacement produces a clean InvalidRequest consistently across
all paths and all key shapes (single, compound PK, composite CK).

The previous approach on this PR branch patched three call sites in
cql3/restrictions/statement_restrictions.cc, which only covered
SELECT, duplicated the check, and placed it mid-restrictions code
(flagged in review). Dropping those changes in favour of the
root-cause fix here.

Un-xfail the tests this fixes:
- test/cqlpy/test_key_length.py: test_insert_65k_pk, test_insert_65k_ck,
  test_where_65k_pk, test_where_65k_ck, test_insert_65k_ck_composite,
  test_insert_total_compound_pk_err, test_insert_total_composite_ck_err.
- test/cqlpy/cassandra_tests/.../insert_test.py: testPKInsertWithValueOver64K,
  testCKInsertWithValueOver64K.
- test/cqlpy/cassandra_tests/.../select_test.py: testPKQueryWithValueOver64K.

test_insert_65k_pk_compound stays xfail: its oversized value gets
rejected by the Python driver's CQL wire-protocol encoder (see
CASSANDRA-19270) before reaching the server, so the fix can't apply.
Updated its reason. testCKQueryWithValueOver64K stays xfail with an
updated reason: Cassandra silently returns empty for an oversized
clustering key in WHERE, while Scylla now throws InvalidRequest - a
deliberate choice mirroring the partition-key case, documented in
the discussion on #10366.

Add three tight-boundary tests (addressing review feedback on the
previous revision) that pin MAX+1 behaviour for SELECT and INSERT of
both partition and clustering keys.

Update test/cluster/dtest/limits_test.py to match the new message
("Key length of \\d+ is longer than maximum of 65535").

fixes #10366
fixes #12247

Co-authored-by: Alexander Turetskiy <someone.tur@gmail.com>

nyh commented on keys/compound.hh:
```diff
@@ -103,7 +104,12 @@ public:
+            // (and client-visible error messages) line up.
+            // Issues #10366 (SELECT) and #12247 (INSERT) both require a
+            // clean InvalidRequest here rather than a generic server error.
+            throw exceptions::invalid_request_exception(format("Key length of {:d} is longer than maximum of {:d}",
```
I was sure serialize_value() was only used on the replica, but if it's used in the coordinator, everything becomes easier.

There's a bit of an ugliness here: keys/compound.hh is a low-level feature, it's not specific to CQL. So it's very strange this low-level code throws CQL-specific exceptions.

So I don't love this solution, but will not disqualify it either.

ptrsmrn replied:
Alternative was to throw a domain-neutral exception (std::invalid_argument or a new keys::key_too_long) and translate it to invalid_request_exception at every CQL boundary that calls serialize_value, but that's many catch sites instead of just this one. Also, there's strong precedent in the codebase for other low-level layers (outside of cql3/) throwing exceptions::invalid_request_exception directly.

nyh commented on test/cqlpy/test_key_length.py:
```diff
@@ -87,6 +83,30 @@ def test_where_65k_ck(cql, table1):
+    stmt = cql.prepare(f'SELECT * FROM {table1} WHERE p = ?')
+    over = 'x' * (65535 + 1)
+    with pytest.raises(InvalidRequest, match='Key length'):
+        cql.execute(stmt, [over])
```
We already have exactly the same test test_insert_65k_pk for 65*1024 - and this test is for 64*1024. I think it's not really necessary to have both... But I guess it's not a disaster.

ptrsmrn replied:
They differ on purpose. test_insert_65k_pk uses 65*1024 = 66560, i.e. clearly over the limit, ~1KB above; it's a smoke test that "very large" is rejected. The new test_*_at_max_pk_plus_one_fails use exactly 65535+1 = 65536 to pin MAX-allowed + 1, the tight boundary. The pair was added in response to @nuivall's earlier review on this PR https://github.com/scylladb/scylladb/pull/23433#discussion_r2012863246, which asked for MAX allowed and MAX allowed + 1 cases. I left the existing 65*1024 tests in place because they reproduce the original bug reports (#10366, #12247) verbatim.

nyh commented on test/cqlpy/test_key_length.py:
```diff
@@ -137,8 +157,12 @@ def table2(cql, test_keyspace):
-@pytest.mark.xfail(reason="Issue #12247")
+# 0 <= number <= 65535". Scylla uses the same CQL wire protocol, so the
+# Python driver itself rejects the oversized component before it ever
+# reaches the server - we therefore cannot produce a clean InvalidRequest
```
I don't understand... The "H format" message came from Cassandra, not from Python... Or am I wrong?
If it's a Python problem, the whole test should be rethought - we don't want an "xfail" when it is a driver bug, not a Scylla bug. But is it a driver bug? I'm not convinced. Maybe you can show some evidence?

ptrsmrn replied:
My understanding is it's a driver issue: python driver computes the token-aware routing key client-side before sending the request. For a compound partition key it serializes each component with a 2-byte length prefix:
`cassandra/query.py:304`:
> yield struct.pack(">H%dsB" % l, l, p, 0)

When l > 65535, struct.pack raises struct.error("'H' format requires 0 <= number <= 65535"), which the driver wraps in a NoHostAvailable. Reproduced by running this very test with --runxfail against my fix:

 > cassandra.cluster.NoHostAvailable: ('Unable to complete the operation against any hosts',
>     {<Host: 127.15.3.1:9042 datacenter1>: error("'H' format requires 0 <= number <= 65535")})

Note the value in the dict is error(...) rather than <Error from server: ...> — i.e. it never reached Scylla. Server log confirms zero requests for the oversized key.

I've tightened the in-code comment to point directly at cassandra/query.py:304 (was vaguely referring to "CQL framing limit" before). The xfail stays — we'd need the driver to either lift this limit or surface it as InvalidRequest for the test to pass.

nyh reviewed (approved):
Looks mostly good, so I tentatively approve, but I did leave some comments/requests so please take a look.

nuivall commented on test/cqlpy/cassandra_tests/validation/operations/insert_invalidate_sized_records_test.py:
```diff
@@ -40,17 +42,17 @@ def testsingleValuePk(cql, test_keyspace):
     with create_table(cql, test_keyspace, "(a blob, b blob, PRIMARY KEY ((a, b)))") as table:
         # sum of columns is too large
-        with pytest.raises(InvalidRequest, match=f'Key length of {len(MEDIUM_BLOB)*2} is longer than maximum of 65535'):
+        with pytest.raises(InvalidRequest, match=r'Key length of \d+ is longer than maximum of 65535'):
```
why change this if it xfails?

nuivall replied:
ok, maybe makes sense to align all tests

ptrsmrn replied:
Updating it now means future un-xfail won't be tripped up by a stale literal.
